staging/src/k8s.io/apiserver/pkg/util/flowcontrol/format/formatting.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package format
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	fcv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// This file provides an easy way to mark a value for printing in full
+// detail IF it is printed but without costing a lot of CPU or memory
+// if the value is NOT printed.  The API Priority and Fairness API
+// objects are formatted into JSON.  The other types of objects here
+// are formatted into golang source.
+
+// Stringer marks the given value for custom formatting by this package.
+type Stringer struct{ val interface{} }
+
+var _ fmt.GoStringer = Stringer{}
+
+// Fmt marks the given value for custom formatting by this package.
+func Fmt(val interface{}) Stringer {
+	return Stringer{val}
+}
+
+// GoString returns a golang source expression
+func (sr Stringer) GoString() string {
+	if sr.val == nil {
+		return "nil"
+	}
+	switch typed := sr.val.(type) {
+	case *fcv1a1.FlowSchema:
+		return ToJSON(*typed)
+	case *fcv1a1.FlowDistinguisherMethod:
+		return ToJSON(*typed)
+	case *fcv1a1.PolicyRulesWithSubjects:
+		return ToJSON(*typed)
+	case *fcv1a1.PriorityLevelConfiguration:
+		return ToJSON(*typed)
+	case *fcv1a1.LimitedPriorityLevelConfiguration:
+		return ToJSON(*typed)
+	case *fcv1a1.QueuingConfiguration:
+		return ToJSON(*typed)
+	case fcv1a1.FlowSchema, fcv1a1.FlowSchemaSpec,
+		fcv1a1.FlowDistinguisherMethod,
+		fcv1a1.PolicyRulesWithSubjects,
+		fcv1a1.Subject,
+		fcv1a1.ResourcePolicyRule,
+		fcv1a1.NonResourcePolicyRule,
+		fcv1a1.PriorityLevelConfiguration,
+		fcv1a1.PriorityLevelConfigurationSpec,
+		fcv1a1.LimitedPriorityLevelConfiguration,
+		fcv1a1.LimitResponse,
+		fcv1a1.QueuingConfiguration:
+		return ToJSON(&sr.val)
```

Comment:
This is pretty confusing: for pointers, we call ToJSON(*val), and for non-pointers, we call ToJSON(&val)? Why? Seems like we either always want a pointer or always don't...

Reply:
Hmm, looks like we can get away with either a pointer or not.

Reply:
No more special cases for pointers.

Reply:
I think you just want sr.val, not &sr.val (which will give a pointer-to-pointer if sr.val was a pointer), but I could be wrong there. A trivial test would make it easy to be certain.